### PR TITLE
Add support for Django 4.2 and Python 3.11

### PR DESCRIPTION
**Pull request by Afani97**
- Update `tox.ini` to support django 4.2
- Update `setup.py` to use latest available required packages
- Update docs to fix tox build
- Update release notes for v2.1.0 
- Remove support for python 3.7

Tox report:
Screenshot 2023-07-18 at 3 31 21 PM

### CODE REVIEW

File: rapidsms/utils/modules.py [hunk 1]
```diff
@@ -176,11 +176,11 @@ def get_class(module, superclass=None):
 
     if len(classes) > 1:
         names = ", ".join([cls.__name__ for cls in classes])
-        raise(AttributeError("Module %s contains multiple %s (%s)." %
+        raise (AttributeError("Module %s contains multiple %s (%s)." %
```
Comment (Afani97) [inline rapidsms/utils/modules.py:179]: Flake8 complaining about spacing

File: tox.ini [hunk 2]
```diff
@@ -24,24 +24,24 @@ basepython =
...
        {[default]deps}
 setenv = {[testenv]setenv}
 commands = django-admin makemigrations --dry-run --check
 
 # Try to mimick Read The Docs' environment (Python 3.7, no explicit Django version)
 [testenv:docs]
-basepython = python3.7
-deps = Sphinx>=1.8,<1.9
+basepython = python3.10
+deps = Sphinx<8.0
```
Comment (Afani97) [inline tox.ini:44]: Sphinx support latest 3 python version: https://www.sphinx-doc.org/en/master/internals/release-process.html#python-version-support-policy

**Review by tobiasmcnulty — COMMENTED** [review]
I noticed some minor changes when running `git ls-files -- '*.py' | xargs django-upgrade --target-version=4.2` (same for 3.2, these must be old):

```diff
diff --git a/rapidsms/admin.py b/rapidsms/admin.py
index a3bea9dc..b93219f2 100644
--- a/rapidsms/admin.py
+++ b/rapidsms/admin.py
@@ -11,10 +11,12 @@ class ConnectionInline(admin.TabularInline):
     extra = 1
 
 
+@admin.register(Contact)
 class ContactAdmin(admin.ModelAdmin):
     inlines = (ConnectionInline,)
 
 
+@admin.register(Connection)
 class ConnectionAdmin(admin.ModelAdmin):
     list_display = ('id', 'backend', 'identity', 'contact')
     list_filter = ('backend',)
@@ -24,5 +26,3 @@ class ConnectionAdmin(admin.ModelAdmin):
 
 admin.site.register(App)
 admin.site.register(Backend)
-admin.site.register(Connection, ConnectionAdmin)
-admin.site.register(Contact, ContactAdmin)
diff --git a/rapidsms/backends/kannel/admin.py b/rapidsms/backends/kannel/admin.py
index ff2eabac..2b2d1d20 100644
--- a/rapidsms/backends/kannel/admin.py
+++ b/rapidsms/backends/kannel/admin.py
@@ -2,6 +2,7 @@ from django.contrib import admin
 from rapidsms.backends.kannel.models import DeliveryReport
 
 
+@admin.register(DeliveryReport)
 class DeliveryReportAdmin(admin.ModelAdmin):
     search_fields = ('text', 'sms_id', 'smsc')
     list_display = ('id', 'message_id', 'date', 'identity', 'status',
@@ -10,4 +11,3 @@ class DeliveryReportAdmin(admin.ModelAdmin):
     ordering = ('-date',)
 
 
-admin.site.register(DeliveryReport, DeliveryReportAdmin)
diff --git a/rapidsms/contrib/messagelog/admin.py b/rapidsms/contrib/messagelog/admin.py
index 14943ac4..12830392 100644
--- a/rapidsms/contrib/messagelog/admin.py
+++ b/rapidsms/contrib/messagelog/admin.py
@@ -5,9 +5,9 @@ from django.contrib import admin
 from .models import Message
 
 
+@admin.register(Message)
 class MessageAdmin(admin.ModelAdmin):
     list_display = ("text", "direction", "who", "date")
     list_filter = ("direction", "date")
 
 
-admin.site.register(Message, MessageAdmin)
diff --git a/rapidsms/router/db/admin.py b/rapidsms/router/db/admin.py
index 235261de..dcce9015 100644
--- a/rapidsms/router/db/admin.py
+++ b/rapidsms/router/db/admin.py
@@ -2,6 +2,7 @@ from django.contrib import admin
 from rapidsms.router.db.models import Message, Transmission
 
 
+@admin.register(Message)
 class MessageAdmin(admin.ModelAdmin):
     search_fields = ('text',)
     list_display = ('id', 'date', 'direction', 'text', 'status', 'updated',
@@ -10,6 +11,7 @@ class MessageAdmin(admin.ModelAdmin):
     ordering = ('-updated',)
 
 
+@admin.register(Transmission)
 class TransmissionAdmin(admin.ModelAdmin):
     list_display = ('id', 'date', 'message', 'status', 'connection', 'updated',
                     'sent', 'delivered')
@@ -19,5 +21,3 @@ class TransmissionAdmin(admin.ModelAdmin):
     search_fields = ('message__text',)
 
 
-admin.site.register(Message, MessageAdmin)
-admin.site.register(Transmission, TransmissionAdmin)
```

**Comment by Afani97** [comment]
> I noticed some minor changes when running `git ls-files -- '*.py' | xargs django-upgrade --target-version=4.2` (same for 3.2, these must be old):
> 
> ```diff
> diff --git a/rapidsms/admin.py b/rapidsms/admin.py
> index a3bea9dc..b93219f2 100644
> --- a/rapidsms/admin.py
> +++ b/rapidsms/admin.py
> @@ -11,10 +11,12 @@ class ConnectionInline(admin.TabularInline):
>      extra = 1
>  
>  
> +@admin.register(Contact)
>  class ContactAdmin(admin.ModelAdmin):
>      inlines = (ConnectionInline,)
>  
>  
> +@admin.register(Connection)
>  class ConnectionAdmin(admin.ModelAdmin):
>      list_display = ('id', 'backend', 'identity', 'contact')
>      list_filter = ('backend',)
> @@ -24,5 +26,3 @@ class ConnectionAdmin(admin.ModelAdmin):
>  
>  admin.site.register(App)
>  admin.site.register(Backend)
> -admin.site.register(Connection, ConnectionAdmin)
> -admin.site.register(Contact, ContactAdmin)
> diff --git a/rapidsms/backends/kannel/admin.py b/rapidsms/backends/kannel/admin.py
> index ff2eabac..2b2d1d20 100644
> --- a/rapidsms/backends/kannel/admin.py
> +++ b/rapidsms/backends/kannel/admin.py
> @@ -2,6 +2,7 @@ from django.contrib import admin
>  from rapidsms.backends.kannel.models import DeliveryReport
>  
>  
> +@admin.register(DeliveryReport)
>  class DeliveryReportAdmin(admin.ModelAdmin):
>      search_fields = ('text', 'sms_id', 'smsc')
>      list_display = ('id', 'message_id', 'date', 'identity', 'status',
> @@ -10,4 +11,3 @@ class DeliveryReportAdmin(admin.ModelAdmin):
>      ordering = ('-date',)
>  
>  
> -admin.site.register(DeliveryReport, DeliveryReportAdmin)
> diff --git a/rapidsms/contrib/messagelog/admin.py b/rapidsms/contrib/messagelog/admin.py
> index 14943ac4..12830392 100644
> --- a/rapidsms/contrib/messagelog/admin.py
> +++ b/rapidsms/contrib/messagelog/admin.py
> @@ -5,9 +5,9 @@ from django.contrib import admin
>  from .models import Message
>  
>  
> +@admin.register(Message)
>  class MessageAdmin(admin.ModelAdmin):
>      list_display = ("text", "direction", "who", "date")
>      list_filter = ("direction", "date")
>  
>  
> -admin.site.register(Message, MessageAdmin)
> diff --git a/rapidsms/router/db/admin.py b/rapidsms/router/db/admin.py
> index 235261de..dcce9015 100644
> --- a/rapidsms/router/db/admin.py
> +++ b/rapidsms/router/db/admin.py
> @@ -2,6 +2,7 @@ from django.contrib import admin
>  from rapidsms.router.db.models import Message, Transmission
>  
>  
> +@admin.register(Message)
>  class MessageAdmin(admin.ModelAdmin):
>      search_fields = ('text',)
>      list_display = ('id', 'date', 'direction', 'text', 'status', 'updated',
> @@ -10,6 +11,7 @@ class MessageAdmin(admin.ModelAdmin):
>      ordering = ('-updated',)
>  
>  
> +@admin.register(Transmission)
>  class TransmissionAdmin(admin.ModelAdmin):
>      list_display = ('id', 'date', 'message', 'status', 'connection', 'updated',
>                      'sent', 'delivered')
> @@ -19,5 +21,3 @@ class TransmissionAdmin(admin.ModelAdmin):
>      search_fields = ('message__text',)
>  
>  
> -admin.site.register(Message, MessageAdmin)
> -admin.site.register(Transmission, TransmissionAdmin)
> ```

Updated.

File: docs/topics/router/index.rst [hunk 3]
```diff
@@ -4,7 +4,7 @@ RapidSMS Routers
 
 .. module:: rapidsms.router
 .. module:: rapidsms.router.base
-.. module:: rapidsms.router.blocking
```
Comment (ronardcaktus) [inline docs/topics/router/index.rst:7]: Just curious, why the change to this line and adding `noindex`?

**Review by ronardcaktus — APPROVED** [review]
Based on my limited understanding of the project, looks 🦩

Result of running `tox`
```
  py38-dj32: OK (11.60=setup[9.92]+cmd[1.68] seconds)
  py38-dj42: OK (7.40=setup[6.08]+cmd[1.32] seconds)
  py39-dj32: OK (9.86=setup[7.91]+cmd[1.95] seconds)
  py39-dj42: OK (8.84=setup[7.09]+cmd[1.75] seconds)
  py310-dj32: OK (7.46=setup[6.02]+cmd[1.44] seconds)
  py310-dj42: OK (6.61=setup[5.46]+cmd[1.15] seconds)
  migrations: OK (6.04=setup[5.51]+cmd[0.53] seconds)
  docs: OK (12.01=setup[7.94]+cmd[4.07] seconds)
  flake8: OK (4.90=setup[4.33]+cmd[0.57] seconds)
  coverage: OK (7.78=setup[5.85]+cmd[1.73,0.19] seconds)
  congratulations :) (82.54 seconds)
```

File: setup.py [hunk 4]
```diff
@@ -16,10 +16,10 @@ def read_file(filename):
     license="BSD",
 
     install_requires=[
```
[the remark this answers is not in the document]
Comment (Afani97) [reply setup.py:18]: Not specifically for supporting versions Django 3.2 and later so I'll revert but yes that makes sense.

File: docs/topics/router/index.rst (hunk 3 above)
Comment (Afani97) [reply docs/topics/router/index.rst:7]: Yeah good question, flake8 was complaining about it being a duplicate.

File: docs/releases/2.2.0.rst [hunk 5]
```diff
@@ -0,0 +1,21 @@
+================================================
+RapidSMS 2.2.0 release notes (under development)
```
Comment (tobiasmcnulty) [inline docs/releases/2.2.0.rst:2]: This file should be a template for the next release (2.1.0 is the current release). Please remove the release notes related to 2.1.0.

File: docs/releases/2.1.0.rst [hunk 6]
```diff
@@ -1,20 +1,21 @@
 ================================================
-RapidSMS 2.1.0 release notes (under development)
+RapidSMS 2.1.0 release notes (current release)
 ================================================
 
 Welcome to RapidSMS 2.1.0! These release notes cover the new features in 2.1.0
 as well as some :ref:`backwards-incompatible-changes-2.1.0` you'll want to be
-aware of when upgrading from RapidSMS 2.0.0.
+aware of when upgrading from RapidSMS 1.1.0.
```
Comment (tobiasmcnulty) [inline docs/releases/2.1.0.rst:7]: `2.0.0` was the correct prior release here.

File: docs/releases/2.1.0.rst [hunk 7]
```diff
@@ -1,20 +1,21 @@
 ================================================
-RapidSMS 2.1.0 release notes (under development)
+RapidSMS 2.1.0 release notes (current release)
```
Comment (tobiasmcnulty) [inline docs/releases/2.1.0.rst:2]: The "(current release)" note will need to be removed from 2.0.0 as well. Refer to the previous release notes docs to see how they look.

**Review by tobiasmcnulty — COMMENTED** [review]
A few comments on the release notes; otherwise, looks good to me!

File: docs/releases/2.1.0.rst [hunk 8]
```diff
@@ -10,11 +10,12 @@ aware of when upgrading from RapidSMS 2.0.0.
...
+ * Added support for Django 4.2
 
  .. _backwards-incompatible-changes-2.1.0:
 
 Backwards-incompatible changes in RapidSMS 2.1.0
 ================================================
 
-*
+* Dropped support for Python 3.7
+* Dropped support for Django 2.2, Django 4.0, and Django 4.1
```
Comment (tobiasmcnulty) [inline docs/releases/2.1.0.rst:21]:
```suggestion
* Dropped support for Django 2.2
```

**Review by tobiasmcnulty — APPROVED** [review]
Looks good and works for me! 🎊 

```
  py38-dj32: OK (6.37=setup[3.97]+cmd[2.39] seconds)
  py39-dj32: OK (5.80=setup[3.39]+cmd[2.42] seconds)
  py310-dj32: OK (5.72=setup[3.30]+cmd[2.42] seconds)
  py38-dj42: OK (5.58=setup[3.27]+cmd[2.30] seconds)
  py39-dj42: OK (5.64=setup[3.35]+cmd[2.29] seconds)
  py310-dj42: OK (5.54=setup[3.30]+cmd[2.25] seconds)
  py311-dj42: OK (5.40=setup[3.27]+cmd[2.13] seconds)
  migrations: OK (3.84=setup[3.32]+cmd[0.53] seconds)
  docs: OK (10.30=setup[3.41]+cmd[6.89] seconds)
  flake8: OK (3.65=setup[3.23]+cmd[0.42] seconds)
  coverage: OK (6.85=setup[3.37]+cmd[3.14,0.34] seconds)
  congratulations :) (64.75 seconds)
```
